supyr_struct/field_type_methods/decoders.py

```diff
@@ -0,0 +1,212 @@
+
+
+def no_decode(self, rawdata, desc=None, parent=None, attr_index=None):
+    ''''''
+    return rawdata
+
+
+def decode_numeric(self, rawdata, desc=None, parent=None, attr_index=None):
+    '''
+    Converts a bytes object into a python int.
+    Decoding is done using struct.unpack
+
+    Returns an int decoded represention of the "rawdata" argument.
+    '''
+    return self.struct_unpacker(rawdata)[0]
+
+
+def decode_decimal(self, rawdata, desc=None, parent=None, attr_index=None):
+    '''
+    Converts a bytes object into a python Decimal.
+
+    Returns a Decimal represention of the "rawdata" argument.
+    '''
+    if self.endian == '<':
+        endian = 'little'
+    else:
+        endian = 'big'
+    d_exp = parent.get_meta('DECIMAL_EXP', attr_index)
+    bigint = str(int.from_bytes(
+        rawdata, endian, signed=self.enc.endswith('S')))
+
+    return Decimal(bigint[:len(bigint)-d_exp] + '.' +
+                   bigint[len(bigint)-d_exp:])
+
+
+def decode_24bit_numeric(self, rawdata, desc=None,
+                         parent=None, attr_index=None):
+    '''
+    Converts a 24-bit bytes object into a python int.
+    Decoding is done using struct.unpack and a manual twos-signed check.
+
+    Returns an int decoded represention of the "rawdata" argument.
+    '''
+    if self.endian == '<':
+        rawint = unpack('<I', rawdata + b'\x00')[0]
+    else:
+        rawint = unpack('>I', b'\x00' + rawdata)[0]
+
+    # if the int can be signed and IS signed then take care of that
+    if rawint & 0x800000 and self.enc[1] == 't':
+        return rawint - 0x1000000  # 0x1000000 == 0x800000 * 2
+    return rawint
+
+
+def decode_timestamp(self, rawdata, desc=None, parent=None, attr_index=None):
+    '''
+    '''
+    return ctime(self.struct_unpacker(rawdata)[0])
+
+
+def decode_string(self, rawdata, desc=None, parent=None, attr_index=None):
+    '''
+    Decodes a bytes object into a python string
+    with the delimiter character sliced off the end.
+    Decoding is done using bytes.decode
+
+    Returns a string decoded represention of the "rawdata" argument.
+    '''
+    return rawdata.decode(encoding=self.enc).split(self.str_delimiter)[0]
+
+
+def decode_raw_string(self, rawdata, desc=None, parent=None, attr_index=None):
+    '''
+    Decodes a bytes object into a python string that can contain delimiters.
+    Decoding is done using bytes.decode
+
+    Returns a string decoded represention of the "rawdata" argument.
+    '''
+    return rawdata.decode(encoding=self.enc)
+
+
+def decode_string_hex(self, rawdata, desc=None, parent=None, attr_index=None):
+    '''
+    Decodes a bytes object into a python string representing
+    the original bytes object in a hexadecimal form.
+
+    Returns a string decoded represention of the "rawdata" argument.
+    '''
+    # slice off the first 2 characters since they are '0x'
+    node = hex(int.from_bytes(rawdata, 'big'))[2:]
+    return '0'*(len(rawdata)*2 - len(node)) + node
```

Comment:
```py
length = len(rawdata) * 2
return (('%%0%dX' % (length)) % (int.from_bytes(rawdata, 'big')))[-length:]
```

The last part `[-length:]` shouldn't even be needed. But there is a "simpler" probably faster way to do this.
Make the X lower or upper case depending on if you want the hex to be upper or lower.

Reply:
I'll try it and implement it if it works out